### ISSUE TITLE
Tracking output

### ISSUE THREAD

**Issue by Vaibhavck** [pytorch/pytorch#40355]
## 🚀 Feature

Tracking the current **dimensions of the output** during **convolution operations** which can be then used to specify the input dimensions for **first** linear layer.
.
## Motivation

I always get frustrated when it comes to specify the input dimensions for the first linear layer when I am using **convolutional layers** before that layer. I need to manually keep track of the dimensions. Or  I need to run the code by putting some random value ( which may not be the expected value) for input to the linear layer and find out the correct value by looking at the error.

## Pitch

Basic idea is to add a feature (a function) which can **keep track of the dimensions** and then use that function, by calling it, to **get the correct input size** ( for the first linear layer ).  So that we don't have to keep track of the dimension ourselves.

## Alternatives

Alternative solution would be to store these dimensions automatically ( similar to the automatic differentiation functionality ).

## Additional context

I am ready to work on this feature for months and gain all the advance knowledge I need.

**Comment by zou3519** [comment]
@Vaibhavck are you talking about a shape-checking API? It sounds like, given a convolution layer, you would like to understand the shape of the outputs given some input shapes and vice versa. Is that correct, and if not, could you provide a concrete example (perhaps with pseudocode) of what you're looking for?

**Comment by Vaibhavck** [comment]
Yes, @zou3519 , that's exactly what I want to do. Here is a screenshot...

<img width="742" alt="pytorch" src="https://user-images.githubusercontent.com/47863940/85498137-c2095900-b5fc-11ea-947a-34242afd73b5.png">

Thank You.

**Comment by ezyang** [comment]
I definitely think a feature like this would be pretty nifty, but there are a bunch of important design questions you'd have to answer first. For example: how does the user interact with this feature? Do they drop a "don't know" token into the module and then run their program and PyTorch tells you what the output should be? (Then it doesn't seem like this is much of an improvement over randomly generating numbers and then reading the error message. Can we make the error message better?) Do they drop a don't know token, but then somehow you can get the needed input size without actually running the operations in question? (Maybe #38490 can help?) Is there some external IDE integration where it can suggest, autocomplete style, what the number should be?

cc @fylux

**Comment by fylux** [comment]
I think that this feature somehow has been already discussed within the idea of checking tensor shapes in general, since from my point of view if you can check tensors you can check layers. In that sense, I would point you to some related discussions https://github.com/pytorch/pytorch/issues/40373 https://github.com/pytorch/pytorch/issues/26889 https://github.com/python/typing/issues/513 [latest](https://groups.google.com/forum/?oldui=1#!topic/python-shape-checkers/tczSyUG0p-4)

My point is that many tensor operations (and layers) can be verified by the typechecker without using any custom static analysis tool. In particular, for most of operations you want to annotate as a type the dimensions of the tensor, so that you can manually annotate something like `x : Tensor[float, [40,100]]` or infer it from the declaration `x = tensor.rand(40,100)`. For that purpose we need variadics and there is already ongoing work to introduce them into Python, e.g. Pyre already supports them and [here](https://github.com/facebook/pyre-check/blob/master/examples/pytorch/sources/linear_regression.py) are some examples of how it looks like. However, convolutions are a different story. For that purpose, we need to be able to do arithmetic on types, leading to signatures like:

`conv (input : Tensor[N, C, H, W], weight : Tensor[O, C / G, h, w], stride : int, padding: int, groups : int = G) -> Tensor[N, O, (H - h + 2 * padding)`

The work on arithmetic is in very early stages, and for PyTorch to rely on Variadics or arithmetic on types they would need to be discussed and potentially added to Python what can take plenty of time. So summarising, we are moving towards that, but support for convolution will come after support for more simple tensor operations and initialise it will be available only on some typecheckers as experimental.

Regarding @ezyang concerns about how the user would interact with this, he would get a type error saying that a function expected for example Tensor[3,10] but got Tensor[4,9]. Since it is part of typechecking, probably IDEs that already provide typechecking support would include it.

Nonetheless, this is just from the point of view of typing, feel free to explore other options that do not rely on types. However, if the goal is to catch this error before running the code there are not many alternatives.

Finally, for typing discussions I would also suggest to take a look at https://github.com/python/typing/issues/513 since this is a good place to centralise discussions that are important not only for PyTorch but for many other frameworks such as TF or NumPy.

**Comment by vadimkantorov** [comment]
A related issue that may be one more important to decide if something like this is implemented: delaying/decoupling parameter initialization from module construction

**Comment by ezyang** [comment]
Referencing #40527. Lazy parameter initialization is an attractive way of fixing this issue too.
